node/core/pvf/src/prepare/worker.rs

```diff
@@ -233,28 +279,75 @@ async fn recv_request(stream: &mut UnixStream) -> io::Result<(Vec<u8>, PathBuf)>
 			"prepare pvf recv_request: non utf-8 artifact path".to_string(),
 		)
 	})?;
-	Ok((code, tmp_file))
+	let preparation_timeout = framed_recv(stream).await?;
+	let preparation_timeout = Duration::decode(&mut &preparation_timeout[..]).map_err(|_| {
+		io::Error::new(
+			io::ErrorKind::Other,
+			"prepare pvf recv_request: failed to decode duration".to_string(),
+		)
+	})?;
+	Ok((code, tmp_file, preparation_timeout))
 }
 
 /// The entrypoint that the spawned prepare worker should start with. The `socket_path` specifies
 /// the path to the socket used to communicate with the host.
 pub fn worker_entrypoint(socket_path: &str) {
 	worker_event_loop("prepare", socket_path, |mut stream| async move {
 		loop {
-			let (code, dest) = recv_request(&mut stream).await?;
+			let (code, dest, preparation_timeout) = recv_request(&mut stream).await?;
 
 			gum::debug!(
 				target: LOG_TARGET,
 				worker_pid = %std::process::id(),
 				"worker: preparing artifact",
 			);
 
+			// Create a static Mutex. We lock it when either thread finishes and set a flag.
+			let mutex = Arc::new(Mutex::new(false));
+
+			let cpu_time_start = ProcessTime::now();
+
+			// Spawn a new thread that runs the CPU time monitor. Continuously wakes up from
+			// sleeping and then either sleeps for the remaining CPU time, or kills the process if
+			// we exceed the CPU timeout.
+			// TODO: Use a threadpool?
```

Comment:
Could use a second opinion on this. My concern is a bunch of prepare jobs completing quickly and there being many waiting threads in the background. I don't see it being that big of a problem though.

Reply:
We have at least two threads here, don't we? If the prepare job finishes before the sleep, the sleep thread will still be sleeping in the background until it wakes up. This is because threads can't safely be killed AFAIK, so we just wait for it to kill itself on wake. But theoretically we could end up with a bunch of sleeping threads. I think it's fine because I don't think we would have too many prepare jobs at once, and also sleeping threads are not really a drain on resources.

Reply:
The sleep time is small, the thread will kill itself quickly. I don't see how a threadpool would help here as the thread exists in a separate temporary process.

Reply:
I agree, just wanted to double-check. Will remove the comment. What do you mean by "separate temporary process"? The child prepare worker?

Reply:
Yes